### PR TITLE
Player upgrade abilities

### PR DESCRIPTION
**Pull request by Guysus**

### CODE REVIEW

File: Source/SteveSloth/Private/MyPlayer.cpp [hunk 1]
```diff
@@ -266,6 +310,37 @@ void AMyPlayer::MoveLeftRight(const FInputActionValue& Value)
 	// Add Animations here with changing of mesh direction
 }
 
+void AMyPlayer::GrapplingHook(const FInputActionValue& Value)
+{
+	if (bIsAimMode && bIsGrapplingHookUnlocked)
+	{
+		if (bDidGrapple == false)
+		{
+			UE_LOG(LogTemp, Warning, TEXT("Grappling Time!!"))
```
Comment (Saretzky) [inline Source/SteveSloth/Private/MyPlayer.cpp:319]: get rid of testing stuff

File: Source/SteveSloth/Private/MyPlayer.cpp [hunk 2]
```diff
@@ -266,6 +310,37 @@ void AMyPlayer::MoveLeftRight(const FInputActionValue& Value)
...
+		}
+
+		bDidGrapple = true;
+		// USE TIMER TO RESET GRAPPLING HOOK
```
Comment (Saretzky) [inline Source/SteveSloth/Private/MyPlayer.cpp:323]: add the timer

**Review by Saretzky — CHANGES_REQUESTED** [review]
according to the trello card this task requires for actual functionality of the mechanics, rather than empty functions

File: Source/SteveSloth/Private/MyPlayer.cpp [hunk 3]
```diff
@@ -143,6 +148,26 @@ void AMyPlayer::AddGrapplingHook()
...
+void AMyPlayer::AddPropeller()
+{
+	bIsPropellerUnlocked = true;
+}
+
+void AMyPlayer::AddMagnetic()
+{
+	bIsMagneticUnlocked = true;
+}
+
```
Comment (Jeff-Moreau) [inline Source/SteveSloth/Private/MyPlayer.cpp:170]: All of these from 146 to 170 should just be a list of upgradeable items that one function will check if the item is unlocked or not

File: Source/SteveSloth/Private/MyPlayer.cpp [hunk 4]
```diff
@@ -232,6 +260,22 @@ void AMyPlayer::SetupPlayerInputComponent(UInputComponent* PlayerInputComponent)
...
+		inputComponent->BindAction(PShovel, ETriggerEvent::Triggered, this, &AMyPlayer::Shovel);
+		inputComponent->BindAction(PShovel, ETriggerEvent::Completed, this, &AMyPlayer::Shovel);
+
+		inputComponent->BindAction(PPropeller, ETriggerEvent::Triggered, this, &AMyPlayer::Propeller);
+		inputComponent->BindAction(PPropeller, ETriggerEvent::Completed, this, &AMyPlayer::Propeller);
+
+		inputComponent->BindAction(PMagnetic, ETriggerEvent::Triggered, this, &AMyPlayer::Magnetic);
+		inputComponent->BindAction(PMagnetic, ETriggerEvent::Completed, this, &AMyPlayer::Magnetic);
+
```
Comment (Jeff-Moreau) [inline Source/SteveSloth/Private/MyPlayer.cpp:278]: we are never going to use all the functions at the same time. So the functions should not have their own input, should just be one input with a switch state inside it that calls the function based off of which upgrade is being used at that time.

File: Source/SteveSloth/Private/MyPlayer.cpp (hunk 1 above)
Comment (Jeff-Moreau) [reply Source/SteveSloth/Private/MyPlayer.cpp:319]: > get rid of testing stuff

agree

File: Source/SteveSloth/Private/MyPlayer.cpp (hunk 2 above)
Comment (Jeff-Moreau) [reply Source/SteveSloth/Private/MyPlayer.cpp:323]: > add the timer

agree

File: Source/SteveSloth/Private/MyPlayer.cpp [hunk 5]
```diff
@@ -266,6 +310,37 @@ void AMyPlayer::MoveLeftRight(const FInputActionValue& Value)
...
+void AMyPlayer::Propeller(const FInputActionValue& Value)
+{
+}
+
+void AMyPlayer::Magnetic(const FInputActionValue& Value)
+{
+}
+
+
```
Comment (Jeff-Moreau) [inline Source/SteveSloth/Private/MyPlayer.cpp:343]: all these functions need to have functionality in them. As stated above these functions should be called from within a switch statement based on which upgrade is currently active.

File: Source/SteveSloth/Public/MyPlayer.h [hunk 6]
```diff
@@ -112,6 +112,24 @@ class STEVESLOTH_API AMyPlayer : public ACharacter
...
+	UInputAction* PClimbingClaw;
+
+	UPROPERTY(EditAnywhere, Category = "Input|Actions")
+	UInputAction* PShovel;
+
+	UPROPERTY(EditAnywhere, Category = "Input|Actions")
+	UInputAction* PPropeller;
+
+	UPROPERTY(EditAnywhere, Category = "Input|Actions")
+	UInputAction* PMagnetic;
```
Comment (Jeff-Moreau) [inline Source/SteveSloth/Public/MyPlayer.h:128]: we are never going to use all the functions at the same time. So the functions should not have their own input, should just be one input with a switch state inside it that calls the function based off of which upgrade is being used at that time.

**Review by Jeff-Moreau — CHANGES_REQUESTED** [review]

**Review by Saretzky — CHANGES_REQUESTED** [review]
there's still a lot that needs to get done, but here are some minor preliminary reviews. right now this is still a barebones skeleton and there needs actual functionality

File: Source/SteveSloth/Public/MyPlayer.h [hunk 7]
```diff
@@ -13,13 +13,13 @@
 // INCLUDES HERE
 #include "MyAmmoData.h"
 #include "CoreMinimal.h"
-#include "MyLevelManager.h"
```
Comment (Jeff-Moreau) [inline Source/SteveSloth/Public/MyPlayer.h:16]: put back in, should not be removed

File: Source/SteveSloth/Public/MyBaseDestructibleObject.h
Comment (Saretzky) [inline Source/SteveSloth/Public/MyBaseDestructibleObject.h:1]: why is this commented out?

File: Source/SteveSloth/Private/MyBaseDestructibleObject.cpp
Comment (Saretzky) [inline Source/SteveSloth/Private/MyBaseDestructibleObject.cpp:1]: Whys is this commented out?

File: Source/SteveSloth/Public/MyPlayer.h [hunk 8]
```diff
@@ -116,50 +126,81 @@ class STEVESLOTH_API AMyPlayer : public ACharacter
...
+	TArray<EAbility*> Abilities;
 	UTexture2D* EquippedAmmoIcon;
 	TArray<UTexture2D*> AmmoIcons;
 	USpringArmComponent* CameraArm;
 	UCameraComponent* PlayerCamera;
 	UEnhancedInputLocalPlayerSubsystem* CurrentIMC;
 	
 	EMappingInputs IMCInputs;
-	AMyLevelManager* LevelManager;
+	EAbility Ability;
```
Comment (Saretzky) [inline Source/SteveSloth/Public/MyPlayer.h:179]: are we using this variable?

File: Source/SteveSloth/Public/MyItemData.h [hunk 9]
```diff
@@ -22,17 +22,15 @@ enum class EItemType
...
-	GreenBud UMETA(DisplayName = "Green Bud"),
-	BlueBud UMETA(DisplayName = "Blue Bud"),
-	RedBud UMETA(DisplayName = "Red Bud"),
-	PebbleAmmo UMETA(DisplayName = "Pebble Ammo"),
-	FirecrackerAmmo UMETA(DisplayName = "Firecracker Ammo"),
-	WaterBalloonAmmo UMETA(DisplayName = "Water Balloon Ammo"),
-	PoisonSacAmmo UMETA(DisplayName = "Poison Sac Ammo"),
-	ChristmasBeetleAmmo UMETA(DisplayName = "Christmas Beetle Ammo"),
-	BigRockAmmo UMETA(DisplayName = "Big Rock Ammo"),
-	GrapplingHook UMETA(DisplayName = "Grappling Hook")
```
Comment (Saretzky) [inline Source/SteveSloth/Public/MyItemData.h:35]: why are we removing these?

File: Source/SteveSloth/Public/MyPlayer.h [hunk 10]
```diff
@@ -116,50 +126,81 @@ class STEVESLOTH_API AMyPlayer : public ACharacter
...
+	FTimerHandle GrappleTimerHandle;
+
+public: // GETTERS/ACCESSORS
+	float GetMaxHealth() { return MaxHealth; }
+	float GetCurrentHealth() { return CurrentHealth; }
+	float GetNeededAmmoIndex();
+
+public: // SETTERS/MUTATORS
+	void SetMaxHealth(float amount) { MaxHealth = MaxHealth + amount; }
+	void SetCurrentHealth(float amount) { CurrentHealth = CurrentHealth + amount; }
```
Comment (Saretzky) [inline Source/SteveSloth/Public/MyPlayer.h:161]: move the getters and setters back to right above the constructor

**Review by Saretzky — CHANGES_REQUESTED** [review]
get rid of print screen in the switch ability (maybe have some sort of UI element that shows what ability)
image

**Comment by Saretzky** [comment]
> get rid of print screen in the switch ability (maybe have some sort of UI element that shows what ability) image

you can use the player hud blueprint

**Review by Jeff-Moreau — CHANGES_REQUESTED** [review]
this is to start.. i need to load your branch

**Review by Jeff-Moreau — CHANGES_REQUESTED** [review]
image
All this needs to be fixed

**Comment by Jeff-Moreau** [comment]
You removed all crates out of the game.. they need to be put back

**Comment by Jeff-Moreau** [comment]
Swimming does not work, the colliders you put in place are missing

File: Source/SteveSloth/Private/MyPlayer.cpp [hunk 11]
```diff
@@ -14,31 +14,34 @@ AMyPlayer::AMyPlayer()
 {
 	PrimaryActorTick.bCanEverTick = true;
 
-	LevelManager = AMyLevelManager::GetInstance();
```
Comment (Jeff-Moreau) [inline Source/SteveSloth/Private/MyPlayer.cpp:17]: put this back it is needed

File: Source/SteveSloth/Private/MyPlayer.cpp [hunk 12]
```diff
@@ -185,26 +185,6 @@ void AMyPlayer::EquipAmmo(EAmmoType ammoType)
...
-		}
-		else
-		{
-			continue;
-		}
-	}
-
-	return neededAmmoIndex;
-}
-
```
Comment (Jeff-Moreau) [inline Source/SteveSloth/Private/MyPlayer.cpp:207]: Whys is this removed?

File: Source/SteveSloth/Public/MyEnemyBaseClass.h [hunk 13]
```diff
@@ -56,16 +56,16 @@ class STEVESLOTH_API AMyEnemyBaseClass : public ACharacter
...
 	bool bIsDead;
 	bool bIsIdle;
+	bool bIsFrozen;
 	bool bIsChasing;
+	bool bIsConfused;
 	bool bIsPatroling;
 	bool bIsAttackingMelee;
+	bool bIsCurrentlyFrozen;
 	bool bIsAttackingRanged;
+	bool bIsCurrentlyConfused;
```
Comment (Jeff-Moreau) [inline Source/SteveSloth/Public/MyEnemyBaseClass.h:68]: I understand what you were doing here.. but you were suppose to be working on the player not organizing other code.. things like this mess up code.. stick to what you were doing..

File: Source/SteveSloth/Public/MyItemData.h (hunk 9 above)
Comment (Jeff-Moreau) [inline Source/SteveSloth/Public/MyItemData.h:35]: This messed up the BP_objects as well they are now missing materials.. again if not needed to touch for your task please dont

File: Source/SteveSloth/Public/MyPlayer.h [hunk 14]
```diff
@@ -116,50 +126,81 @@ class STEVESLOTH_API AMyPlayer : public ACharacter
 	UPROPERTY(EditAnywhere, Category = "Input|Actions")
 	UInputAction* PAiming;
 
+	UPROPERTY(EditAnywhere, Category = "Input|Actions")
+	UInputAction* PSwitchAbilities;
+
+	UPROPERTY(EditAnywhere, Category = "Input|Actions")
+	UInputAction* PCamPitch;
```
Comment (Jeff-Moreau) [inline Source/SteveSloth/Public/MyPlayer.h:133]: There was already a pitch input, what was the purpose of adding another?

File: Source/SteveSloth/Public/MyPlayer.h [hunk 15]
```diff
@@ -116,50 +126,81 @@ class STEVESLOTH_API AMyPlayer : public ACharacter
...
 
+	// You can expose some of your collision query data as properties to help customize and debug 
+	// Here we expose the collision channel we want to run the query on, and set it to only hit Pawns.
+	UPROPERTY(EditAnywhere, Category = "Collision")
+	TEnumAsByte<ECollisionChannel> TraceChannelProperty = ECC_Pawn;
+
+
+	FVector TraceEnd;
+	FVector TraceStart;
+	FVector WallHitLocation;
```
Comment (Jeff-Moreau) [inline Source/SteveSloth/Public/MyPlayer.h:149]: I dont see where these three variables are used.. if they are cool, if not please remove

**Review by Jeff-Moreau — CHANGES_REQUESTED** [review]

**Review by Saretzky — APPROVED** [review]
looks good

**Pull request merged**